File: core/src/main/scala/com/graphbrain/eco/nodes/CondsFun.scala
package com.graphbrain.eco.nodes

import com.graphbrain.eco.NodeType.NodeType
import com.graphbrain.eco.{Contexts, NodeType}

class CondsFun(params: Array[ProgNode], lastTokenPos: Int= -1) extends FunNode(params, lastTokenPos) {
  override val label = ";"

  override def ntype: NodeType = NodeType.Boolean

  override def booleanValue(ctxts: Contexts) = {
    for (p <- params) {
      p.value(ctxts)
      p.ntype match {
        case NodeType.Boolean => {
          for (c <- ctxts.ctxts) {
            if (!c.getRetBoolean(p)) {
              ctxts.remContext(c)
            }
          }
          ctxts.applyChanges()
        }
      }
    }
  }

  override protected def typeError() = error("parameters must be boolean")
}

File: core/src/main/java/com/graphbrain/eco/nodes/CondsFun.java
package com.graphbrain.eco.nodes;

import com.graphbrain.eco.Context;
import com.graphbrain.eco.NodeType;
import com.graphbrain.eco.Contexts;

public class CondsFun extends FunNode {

    public CondsFun(ProgNode[] params, int lastTokenPos) {
        super(params, lastTokenPos);
    }

    public CondsFun(ProgNode[] params) {
        this(params, -1);
    }

    @Override
    public String label(){return ";";}

    @Override
    public NodeType ntype(){return NodeType.Boolean;}

    @Override
    public void booleanValue(Contexts ctxts) {
        for (ProgNode p : params) {
            p.value(ctxts);
            if (p.ntype() == NodeType.Boolean) {
                for (Context c : ctxts.getCtxts()) {
                    if (!c.getRetBoolean(p)) {
                        ctxts.remContext(c);
                    }
                }
                ctxts.applyChanges();
            }
        }
    }
}